source/blog/2016/git-git-the-world-s-smallest-git-plugin.html.markdown

```diff
@@ -0,0 +1,157 @@
+$ git git rebase
+git: 'git' is not a git command. See 'git --help'.
+
+Did you mean this?
+	init
+```
+
+Or sometimes I will be inside a [gitsh] shell -- where the `git` prefix is
+implied -- but still type `git` out of habit[^1]. Oops!
+
+  [gitsh]: https://github.com/thoughtbot/gitsh
+
+I've been putting up with this for years, grumbling to myself, "No, I did not
+mean init..." and a few weeks ago decided to do something about it.
+
+The solution turned out to be hiding in the error message I had read so many
+times: `'git' is not a git command`. What if it *was* a git command?
+
+## Git Commands
+
+What actually happens when you run `git status`? For a long time, I imagined
+the `git` command looked something like this:
+
+```ruby
+subcommand = ARGUMENTS.first
+arguments = ARGUMENTS.rest
+
+case subcommand
+when "status": git_status(arguments)
+when "commit": git_commit(arguments)
+# ...
+else
+  abort("'#{subcommand}' is not a git command")
+end
+```
+
+There are many variations on this style, but it's my default method for
+building command-line tools with subcommands.
+
+But Git actually takes a very different approach, motivated by the desire to
+make extensions feel natural to use and easy to implement.
+
+Each Git command is implemented as a standalone command named by convention
+`git-COMMAND`[^2]. This includes builtin commands like `git status`. You can see
+all of them by looking in the `libexec/git-core` directory included with your
+Git installation. For me, that's:
+
+```
+$ /usr/local/Cellar/git/2.9.0/libexec/git-core
+
+git-add
+# ...
+git-blame
+# ...
+git-checkout
+# ...
+git-status
+# ... 166 commands in total
+```
+
+So the `git` command actually looks more like this:
+
+```ruby
+subcommand = ARGUMENTS.first
+arguments = ARGUMENTS.rest
+
+if command_exists?("git-#{subcommand}")
+  exec("git-#{subcommand}", arguments)
+else
+  abort("'#{subcommand}' is not a git command")
+end
+```
+
+When Git checks if the command exists, it looks in the `libexec` directory. But
+it also searches for commands on your `$PATH`, which is how third-party plugins
+can extend git's behavior.
+
+## `git-git`
+
+Now that we know how git finds commands, we can implement our new `git`
+subcommand. Based on the conventions outlined above, that means we need a new
+command called `git-git` on our path.
+
+If you don't already have a place you store your own terminal commands, a good
+convention is `$HOME/bin`:
+
+```
+$ mkdir -p $HOME/bin
+$ export PATH="$HOME/bin:$PATH"
+```
+
+Then we'll need to create the `git-git` command and make it executable:
+
+```
+$ touch $HOME/bin/git-git
+$ chmod +x $HOME/bin/git-git
+```
+
+If we run `git git status` we no longer see an error!
+
+Now we can implement the script. When we run `git git status`, our `git-git`
+command is run with the single argument `status`. So to run the command we
+intended, we can re-execute `git` with the arguments to the script. Here's what
+`git-git` looks like:
+
+```
+exec git "$@"
```

Comment:
I've dropped the `exec` from the blog post, though I'll keep it in my actual script. I think it's more correct with it there -- for reporting exit codes, etc. -- but out of scope to explain in this post.